### PR TITLE
RFC 0009 -- Code Review.

### LINKED ISSUE #21: Create RFC around Github and PR etiquette
How do authors and reviewers interact with PRs? Document the standard process that we typically have used here at CDS.
(Usually authors merge, review required before merging, whether we prefer squash and merge, whether or when we want design approval or when just dev is enough, etc etc). I have my own opinions about a lot of these, based on how I've operated since being at CDS, but I am sure these would be better if discussed and agreed upon by other folks.

Also put in guidelines for how reviewers and reviewees should act towards each other in a PR review. I know paul had some documentation around this that was excellent. I will try to find it.

### PR DESCRIPTION
**Pull request by dj2**
# Summary | Résumé

> 1-3 sentence description of the changed you're proposing, including a link to
This adds an RFC around code reviews at CDS. Provides some
considerations and guide rails for how they should work.

This is based primarily on
https://github.com/cds-snc/cra-claim-tax-benefits/blob/main/docs/CODE-REVIEW-GUIDELINES.md
by @pcraig3 with suggestions and contributions from @brdunfield.

Fixes #21.

### CODE REVIEW

File: rfcs/0009-code-review.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+subjectivity by giving guide rails to make reviews more consistent.
+
+## Proposal
+Below is a list of considerations these are divided into principles and
+technical consideration. Focus on the principles and then lead into the
+technical considerations.
+
+### General guidelines
+ - Code reviews are required at CDS.
+   - Each review should have a technical reviewre.
```
[the remark this answers is not in the document]
Comment (dj2) [reply rfcs/0009-code-review.md:35]: Can I pretend it's the British spelling?

[the remark this answers is not in the document]
Comment (dj2) [reply rfcs/0009-code-review.md]: Fixed.

File: rfcs/0009-code-review.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+     reviwere would write it. Just, good.
+ - In general authors should merge their own PRs.
+   - This allows them to make any final touch ups or edit the commit
+     message.
+ - Cleanup branches after merge. If possible, enable the auto-delete
+   after merge option on repositories.
+ - Prefer to `Squash and merge` commits.
+   - The development history is in the PR if desired, but typically
+     doesn't add much to the main repository.
+
```
Comment (maxneuvians) [inline rfcs/0009-code-review.md:56]: Something to add: No force pushes by admin

File: rfcs/0009-code-review.md [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  - practices
+---
+
+[La version française suit.](#résumé)
+
+# Code Review
+
+## Summary
+Code review is an integral part of how we work at CDS. Reviewing code
+provides an opportunity for other developers to understand the code. It
```
Comment (brdunfield) [inline rfcs/0009-code-review.md:17]: Nitpicky but "code" is used a lot in this sentence / paragraph. Maybe just "to understand it" here?

File: rfcs/0009-code-review.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+   - The reviewer may not understand all the considerations a reviewee
+     put into the code.
+   - The reviewee may not understand the reviewers past experience with
+     the code.
+ - [Be nice](https://qz.com/work/625870/after-years-of-intensive-analysis-google-discovers-the-key-to-good-teamwork-is-being-nice/)
+   - Read the CDS [Code of Conduct](https://github.com/cds-snc/.github/blob/master/CODE_OF_CONDUCT.md)
+ - Escalate if the review is blocked.
+   - The goal of the review is that the code looks good to the reviewer.
+     It doesn't have to be perfect, it doesn't have to be how the
+     reviwere would write it. Just, good.
```
Comment (brdunfield) [inline rfcs/0009-code-review.md:47]: typo "reviwere"

File: rfcs/0009-code-review.md (hunk 2 above)
Comment (dj2) [reply rfcs/0009-code-review.md:56]: I'm hesitant to add that. Sometimes a to-be-reviewed commit is needed. They should be rare, but I didn't want to blanket deny them. Maybe something like:

- Force pushing to a repo should be a last resort.

File: rfcs/0009-code-review.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+   - Marking the comments as `Resolved` makes it easier to understand
+     what still needs to be done.
+ - Assume the best intention from review comments
+ - Wait for the tests to pass before you merge.
+ - Tend towards smaller changes. They are easier to understand and
+   faster to review.
+
+
+### Principles for reviewers
+ - Try to comment on the PR in as short an interval is possible (<24
```
Comment (bryan-robitaille) [inline rfcs/0009-code-review.md:85]: 24 Business hours is essentially a little over 3 business days.  I would wonder if this timeframe should be shortened especially if the comment isn't necessarily a review but when they'll be able to review.

File: rfcs/0009-code-review.md (hunk 4 above)
Comment (dj2) [reply rfcs/0009-code-review.md:47]: Weird, VIM thinks that one is spelt correctly. Will fix.

File: rfcs/0009-code-review.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+#### Tests
+ - Do tests exist, and are they comprehensive, covering good input, bad
+   input and errors?
+ - Are tests checking at the right level?
+ - Do the existing set of tests pass?
+
+#### Documentation
+ - Does this change require an update to the README or configuration
+   files?
```
Comment (brdunfield) [inline rfcs/0009-code-review.md:138]: I might add any non-github documentation here as well. Some internal facing docs are stored in Drive and often just get left to rot as the codebase changes.

File: rfcs/0009-code-review.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+ - Is the reason for the change outlined in the commit message?
+
+
+## Drawbacks
+Making the code review process too restrictive makes it more of a core
```
Comment (brdunfield) [inline rfcs/0009-code-review.md:143]: Should this be "chore"?

File: rfcs/0009-code-review.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+then it should be. We want the process to be fairly light, but still
+maintain value.
+
+
+## Alternatives
+Without the guide rails the process can continue as is with different
+styles of code review being provided by different folks. This, while
+functional, can be difficult if you need different reviewers for things
```
Comment (brdunfield) [inline rfcs/0009-code-review.md:151]: I think this sentence might be missing a couple words or commas? I'm finding it difficult to read and understand what it's trying to say 🤔

File: rfcs/0009-code-review.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+     put into the code.
+   - The reviewee may not understand the reviewers past experience with
+     the code.
+ - [Be nice](https://qz.com/work/625870/after-years-of-intensive-analysis-google-discovers-the-key-to-good-teamwork-is-being-nice/)
+   - Read the CDS [Code of Conduct](https://github.com/cds-snc/.github/blob/master/CODE_OF_CONDUCT.md)
+ - Escalate if the review is blocked.
+   - The goal of the review is that the code looks good to the reviewer.
+     It doesn't have to be perfect, it doesn't have to be how the
+     reviwere would write it. Just, good.
+ - In general authors should merge their own PRs.
```
Comment (bryan-robitaille) [inline rfcs/0009-code-review.md:54]: I would also highlight that if someone else but the reviewee presses the 'squash and merge' on the PR the reviewee loses all credit to their contribution and the merged commit is in the name of the reviewer (or person who pressed 'squash and merge'.

File: rfcs/0009-code-review.md (hunk 3 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Rewrote the whole paragraph, better?

File: rfcs/0009-code-review.md (hunk 4 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: done.

File: rfcs/0009-code-review.md (hunk 2 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: done, seem good?

File: rfcs/0009-code-review.md (hunk 6 above)
Comment (dj2) [reply rfcs/0009-code-review.md:144]: Added a sub-bullet, good?

File: rfcs/0009-code-review.md (hunk 7 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Good catch. Fixed.

File: rfcs/0009-code-review.md (hunk 8 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Hah, missed the forest because of the trees. Better?

File: rfcs/0009-code-review.md (hunk 3 above)
Comment (brdunfield) [reply rfcs/0009-code-review.md]: Very nice!

**Review by brdunfield — APPROVED** [review]
LGTM! Thanks for the changes 🦑

File: rfcs/0009-code-review.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+ - Programming is subjective.
+   - Different approaches are valid.
+   - Ask for clarifications if you need any.
+ - Communicate which ideas you feel strongly about and those you don't.
+
+
+### Technical Considerations
+
+#### Logic
+ - Does the code work?
```
Comment (bryan-robitaille) [inline rfcs/0009-code-review.md:106]: - Are there tests that were written to cover the new logic?
- Does this introduce a certain level of tech debt?
- If there is an automated spun up instance for the PR has the functionality introduced in the PR been tested through the UI.

File: rfcs/0009-code-review.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+   - Add Design, Research, Product Management, Service Owner, as needed.
+ - Tend towards questions rather then statements.
+   - The reviewer may not understand all the considerations a reviewee
+     put into the code.
+   - The reviewee may not understand the reviewers past experience with
+     the code.
+ - [Be nice](https://qz.com/work/625870/after-years-of-intensive-analysis-google-discovers-the-key-to-good-teamwork-is-being-nice/)
+   - Read the CDS [Code of Conduct](https://github.com/cds-snc/.github/blob/master/CODE_OF_CONDUCT.md)
+ - Escalate if the review is blocked.
+   - The goal of the review is that the code looks good to the reviewer.
```
Comment (bryan-robitaille) [inline rfcs/0009-code-review.md:51]: What is 'good'?  Wondering if we need to have clearer common understanding.  For example:
- Is 'good' tied to the reviewers underlying knowledge of the programming language used and/or the project itself?
- Is a 'hacky' solution good because it meets an immediate need vs introducing tech debt and future work refactoring?

**Review by bryan-robitaille — COMMENTED** [review]
I think this is a great start.  Added some comments to hopefully spark discussion around 'what is good' and things to check when reviewing a PR.  I think that checking the new functionality or changes in a running instance is an important one.  I've been caught out many times by purely reviewing the 'code' only to miss bugs that were being introduced because I didn't check a running instance of the PR branch.

File: rfcs/0009-code-review.md (hunk 5 above)
Comment (dj2) [reply rfcs/0009-code-review.md:85]: I phrased that badly. I was trying to say 24 hrs, but excluding weekends. I'll try to rephrase so it's correct.

File: rfcs/0009-code-review.md (hunk 11 above)
Comment (dj2) [reply rfcs/0009-code-review.md:51]: That's a really good question and .... I think it depends?

To me, the answer is very context dependant. Sometimes it's worthwhile taking on tech debt to accomplish some goal. You need to know you're taking on the debt and have some plans around containing it, but taking it on maybe the right call. So, a hacky solution maybe the correct solution for now.

So, I think for this 'good' means in the context of the timelines, goals and constraints of the PR.  It's also tied to the reviewers abilities. My ability to review 'good' python code is small, I can do it but I'm not a python expert (or intermediate) so my good is a lot different then other folks. My good would be understanding as a non-python dev, comments, structure, but I'd probably also request a python person review the code as well.

Thoughts?

File: rfcs/0009-code-review.md (hunk 9 above)
Comment (dj2) [reply rfcs/0009-code-review.md:54]: Good idea, added. Better?

File: rfcs/0009-code-review.md (hunk 5 above)
Comment (dj2) [reply rfcs/0009-code-review.md:87]: Better?

File: rfcs/0009-code-review.md (hunk 10 above)
Comment (dj2) [reply rfcs/0009-code-review.md:108]: Added the automated instance point to the tests section below, I think the are there tests is covered there?

Added the tech debt part here, that's a good addition.

**Comment by dj2** [comment]
Thanks a lot for the suggestions @bryan-robitaille I think I've integrated most of it? Was there a broader change you'd like to see around testing in the UI or is it covered by the current additions?

File: rfcs/0009-code-review.md (hunk 11 above)
Comment (wmoussa-cds) [reply rfcs/0009-code-review.md:51]: I agree the sentence is too ambiguous. Could we use the Technical Considerations as a north star goal for any code review? For e.g. _The goal of the review is that the reviewers ensure the code follows the technical considerations defined in this RFC._

File: rfcs/0009-code-review.md [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+ - Does the code work?
+   - Does it perform its intended function?
+   - Is the logic correct?
+   - Is the proposed UI accessible?
+ - Does this the change introduce tech debt?
+   - Adding tech debt isn't a bad thing, but we need to make sure we understand
+     why we're adding it and how we're going to contain it.
+   - If adding, are there comments or documentation around the debt so future
+     maintainers can understand the issues?
+ - Does this change introduce any security concerns?
```
Comment (wmoussa-cds) [inline rfcs/0009-code-review.md:117]: should we ask if it impacts any existing ATO/Security paperworks?

File: rfcs/0009-code-review.md [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+   intended?
+ - Are potential errors being caught and logged?
+ - Are third-party utilities used which may return errors?
+ - Is any unusual behavior or edge-case handling described?
+
+#### Tests
+ - Do tests exist, and are they comprehensive, covering good input, bad
+   input and errors?
+ - Are tests checking at the right level?
+ - Do the existing set of tests pass?
```
Comment (wmoussa-cds) [inline rfcs/0009-code-review.md:147]: i think this will be automatically checked by Github Actions

File: rfcs/0009-code-review.md [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+ - If there is an instance created for the PR has the functionality been tested
+   through the UI?
+
```
Comment (wmoussa-cds) [inline rfcs/0009-code-review.md:150]: feels like we're missing something about code formatting and styling? in fact we might need an RFC about that too

File: rfcs/0009-code-review.md [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+
+#### Documentation
+ - Does this change require an update to the README or configuration
+   files?
+   - Are there internal documents which need to be updated?
+ - Is the reason for the change outlined in the commit message?
+
```
Comment (wmoussa-cds) [inline rfcs/0009-code-review.md:157]: what do we do with https://github.com/cds-snc/.github/blob/main/pull_request_template.md ?

Comment (dj2) [reply rfcs/0009-code-review.md:157]: How do folks feel about that template? I think in almost all cases I erase all of it but the initial title and just paste my description under that. I've see it used in other cases, but do people use the checklist and check things off? Do they typically have content in each of the sections?

File: rfcs/0009-code-review.md (hunk 13 above)
Comment (dj2) [reply rfcs/0009-code-review.md:147]: In general yes, so this will probably be a no-op, but not every project has those actions setup, and there are also possibilities that not all tests run on actions.

File: rfcs/0009-code-review.md (hunk 9 above)
Comment (bryan-robitaille) [reply rfcs/0009-code-review.md:54]: Thanks :)

File: rfcs/0009-code-review.md (hunk 5 above)
Comment (bryan-robitaille) [reply rfcs/0009-code-review.md:87]: :)

**Comment by bryan-robitaille** [comment]
Thanks @dj2, I think testing by spinning up the project and testing the newly added functionality is covered off in the new additions.

File: rfcs/0009-code-review.md (hunk 15 above)
Comment (wmoussa-cds) [reply rfcs/0009-code-review.md]: lets ask in the dev community meeting today

File: rfcs/0009-code-review.md (hunk 11 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Better?

File: rfcs/0009-code-review.md (hunk 12 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Done.

File: rfcs/0009-code-review.md (hunk 14 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Better?

File: rfcs/0009-code-review.md (hunk 15 above)
Comment (dj2) [reply rfcs/0009-code-review.md]: Resolving this, will do another PR which updates the template to hopefully make it shorter.

**Comment by wmoussa-cds** [comment]
lgtm

**Review by wmoussa-cds — APPROVED** [review]
